diffkemp/snapshot.py

```diff
@@ -0,0 +1,226 @@
+"""
+Kernel snapshot representation.
+Contains mappings of original kernel source, snapshot kernel source and
+holds the function list.
+"""
+from diffkemp.llvm_ir.kernel_module import LlvmKernelModule
+from diffkemp.llvm_ir.kernel_source import KernelSource
+import os
+import pkg_resources
+import shutil
+import yaml
+
+
+class Snapshot:
+    """
+    Kernel snapshot.
+    Provides methods for snapshot management. Also includes descriptions of
+    functions and sysctl options.
+    """
+
+    class FunctionDesc:
+        """
+        Function description.
+        Contains LLVM module of the function definition, a global variable,
+        and a tag.
+        For sysctl option, glob_var is the data variable that the option sets.
+        """
+
+        def __init__(self, mod, glob_var, tag):
+            self.mod = mod
+            self.glob_var = glob_var
+            self.tag = tag
+
+    class FunctionGroup:
+        """
+        Group of functions. Used to group functions belonging to a single
+        sysctl options.
+        """
+
+        def __init__(self):
+            self.functions = dict()
+
+    def __init__(self, kernel_source=None, snapshot_source=None,
+                 fun_kind=None):
+        self.kernel_source = kernel_source
+        self.snapshot_source = snapshot_source
+        self.fun_kind = fun_kind
+        self.fun_groups = dict()
+
+    def initialize_write(self, output_dir, sysctl=None, setup_dir=True):
+        """
+        Prepares a snapshot for kernel analysation and its directory creation.
+        :param output_dir: Snapshot output directory.
+        :param sysctl: Whether the sysctl option is used.
+        :param setup_dir: Whether to recreate the output directory.
+        """
+        output_path = os.path.abspath(output_dir)
+
+        # Cleanup or create the output directory of the new snapshot.
+        if setup_dir:
+            if os.path.isdir(output_path):
+                shutil.rmtree(output_path)
+            os.mkdir(output_path)
+
+        # Prepare the source representation of the new snapshot
+        self.snapshot_source = KernelSource(output_path)
+
+        # Prepare a new function list based on sysctl value.
+        self.fun_kind = "sysctl" if sysctl else None
```

Comment:
This is fine in my opinion. The reason for this is simple: the initialization inside `self.__init__` is there only to define `self.fun_kind`. The important part is the initialization of write and reading modes, which may so to speak overwrite the unused default value as they can be parameterized. I at least believe that in this way it is better to parameterize the read method and not the constructor so that the two are more clearly differentiated.

Reply:
In that case you can remove `snapshot_source` and `fun_kind` from the constructor's parameters (and set them to `None` for the purpose of defining the attributes), since they are both set in the initialize methods and the arguments in the constructor are unused (and in practice unusable) anyway (and perhaps even move the initialization of `kernel_source` to `initialize_write`).